src/Covenant/ASG.hs

```diff
@@ -444,6 +538,97 @@ ledgerDestruct ::
+  DRepCredential -> ATyLedger (ALedgerData Credential) -- Credential
+  ChangedParameters -> ATyExpr TyPlutusData -- BuiltinData
+  PubKeyHash -> ATyExpr TyByteString -- BuiltinByteString
+  Lovelace -> ATyExpr TyInteger -- Integer
+  CurrencySymbol -> ATyExpr TyByteString -- BuiltinByteString
+  TokenName -> ATyExpr TyByteString -- BuiltinByteString
+  Value -> ATyExpr TyPlutusData -- Map CurrencySymbol (Map TokenName Integer) **
+  ScriptHash -> ATyExpr TyByteString -- BuiltinByteString
+  DatumHash -> ATyExpr TyByteString -- BuiltinByteString
+  Datum -> ATyExpr TyPlutusData -- BuiltinData
+  Redeemer -> ATyExpr TyPlutusData -- BuiltinData
+  Constitution -> ATyExpr TyPlutusData -- Maybe ScriptHash **
+  POSIXTime -> ATyExpr TyInteger -- Integer
+  TxId -> ATyExpr TyByteString -- BuiltinByteString
+  MintValue -> ATyExpr TyPlutusData -- Map CurrencySymbol (Map TokenName Integer) **
+
+typeLedgerFieldAccess :: LedgerRecord -> LedgerFieldName -> Either TypeError TyASGNode
+typeLedgerFieldAccess = \case
+  GovernanceActionId -> \case
+    GaidTxId -> Right $ ATyLedger (ALedgerNewtype TxId) -- TxId
+    GaidGovActionIx -> Right $ ATyExpr TyInteger -- Integer
+    field -> Left $ TyErrInvalidField GovernanceActionId field
+  Committee -> \case
+    CommitteeMembers -> Right $ ATyExpr TyPlutusData -- Map ColdCommitteeCredential Integer **
+    CommitteeQuorum -> Right $ ATyExpr TyPlutusData -- Rational
+    field -> Left $ TyErrInvalidField Committee field
+  ProtocolVersion -> \case
+    PvMajor -> Right $ ATyExpr TyInteger -- Integer
+    PvMinor -> Right $ ATyExpr TyInteger -- Integer
+    field -> Left $ TyErrInvalidField ProtocolVersion field
+  ProposalProcedure -> \case
+    PpDeposit -> Right $ ATyLedger (ALedgerNewtype Lovelace) -- Lovelace
+    PpReturnAddr -> Right $ ATyLedger (ALedgerData Credential) -- Credential
+    PpGovernanceAction -> Right $ ATyLedger (ALedgerData GovernanceAction) -- GovernanceAction
+    field -> Left $ TyErrInvalidField ProposalProcedure field
+  TxInInfo -> \case
+    TxInInfoOutRef -> Right $ ATyLedger (ALedgerRecord TxOutRef) --  TxOutRef
+    TxInInfoResolved -> Right $ ATyLedger (ALedgerRecord TxOut) -- TxOut
+    field -> Left $ TyErrInvalidField TxInInfo field
+  TxInfo -> \case
+    TxInfoInputs -> Right $ listOf (ATyLedger (ALedgerRecord TxInInfo)) -- [TxInInfo] **
+    TxInfoReferenceInputs -> Right $ listOf (ATyLedger (ALedgerRecord TxInInfo)) -- [TxInInfo] **
+    TxInfoOutputs -> Right $ listOf (ATyLedger (ALedgerRecord TxOut)) -- [TxOut] **
+    TxInfoFee -> Right $ ATyLedger (ALedgerNewtype Lovelace) -- Lovelace
+    TxInfoMint -> Right $ ATyLedger (ALedgerNewtype MintValue) -- MintValue
+    TxInfoTxCerts -> Right $ listOf (ATyLedger (ALedgerData TxCert)) -- [TxCert] **
+    TxInfoWdrl -> Right $ ATyExpr TyPlutusData -- Map Credential Lovelace **
+    TxInfoValidRange -> Right $ ATyLedger (ALedgerRecord (Interval $ ALedgerNewtype POSIXTime)) -- type POSIXTimeRange = Interval POSIXTime **
```

Comment:
I've parametrized `Interval a` as `Interval TyLedger`. 
`Interval TyASGNode` is probably more appropriate, but it's causing circular import errors. 
Since we are never using `Interval` with an arbitrary type, I think it's fine to have this restriction for now.